### DISCUSSION TITLE
Session migration does not work with user UUID id

### DISCUSSION THREAD

**Discussion by Miosame** [laravel/framework#35226] [Ideas]
- Laravel Version: 8.7.1
- PHP Version: 7.4.11
- Database Driver & Version: mysql - latest

### Description:

I wanted to report this first before making a PR to the docs, to understand if my proposed solution is not in line with how it should be done or if this is an upstream laravel framework issue as how foreignId is handled?

The issue me and others keep coming across is:

- Change user ID in user migration to be UUID
- now sessions won't work unless you change:
`$table->foreignId('user_id')->nullable()->index();` to `$table->uuid('user_id')->nullable()->index();`

Now the question remains: is this change not how it should be proposed into the docs and should sessions be handled differently to accommodate for both numeric and string user IDs?

Also I'm not sure to which docs to submit this warning/notice to, since it automatically affects multiple regions:

- login/logout not working (due to csrf)
- all csrf tokens failing (mismatch)
- all forms failing (csrf mismatch)

Thanks!

**Comment by ctf0** [comment]
same here as well, i keep getting `419 | page expired` unless `VerifyCsrfToken` mw is disabled

**Comment by Slayking** [comment]
Same here, this has been giving me a headache all day. 

The CSRF token is always mismatched whilst using UUID's. It works fine if you use the File driver but that's not ideal.

**Comment by aibarra11** [comment]
thanks for bringing this up. all the other 419 fixes didn't work. For now, this is working for me.

**Comment by madalinignisca** [comment]
If you have a foreign id to an uuid, always the foreign id must be an uuid.

**Comment by sneakylenny** [comment]
I just ran into this issue and it noticed I forgot to add
```php
protected $keyType = 'string';
```
to my User model. Which resulted in the session setting the first few digits of the uuid as user_id on the session.
After adding the keyType prop it worked and the full uuid was added to the session's user_id.

**Comment by pedrouzcategui** [comment]
Not sure if someone is still having this issue, but the workaround for me was to modify the "AuthenticatedSessionController.php" to manually set the user_id before the session is regenerated, this made it work.

![image](https://github.com/user-attachments/assets/6f7abb06-5ad3-414a-ba48-031bbdc9f09f)

**Comment by sfreytag** [comment]
We just ran into this as well.

An underlying problem seems to be that the exception that might get thrown from a mismatched schema is quietly squashed in the DatabaseSessionDriver class.

After a login the DatabaseSessionDriver tries to write the new session. It detects the session does not exist, so calls `performInsert`. This wraps the query in a try/catch, which catches a `QueryException` and falls back to calling `performUpdate` instead. Presumably this is to deal with the case that session.id exists?

However in our case the schema was out of sync and the QueryException was:

`Illuminate\Database\QueryException: SQLSTATE[01000]: Warning: 1265 Data truncated for column 'user_id'`

This got quietly squashed, the performUpdate also failed (because there was nothing to update) and it took us a lot of debugging to work out the problem. We were not helped by shift to user UUIDs and the shift to a database session driver being over a year apart and it was otherwise working fine with the file driver.

Relevant code:

https://github.com/laravel/framework/blob/12.x/src/Illuminate/Session/DatabaseSessionHandler.php#L156-L160

My suggestion would be - can this catch be made tighter to only catch specific things and otherwise rethrow exceptions that it would be good to know about, particularly those arising from schema changes?

To answer that, it would be good to know more about the specific things this catch was intended to catch.
